cranelift/codegen/src/machinst/valueregs.rs

```diff
@@ -0,0 +1,190 @@
+    parts: [R; VALUE_REGS_PARTS],
+}
+
+/// A type with an "invalid" sentinel value.
+pub trait InvalidSentinel: Copy + Eq {
+    /// The invalid sentinel value.
+    fn invalid_sentinel() -> Self;
+    /// Is this the invalid sentinel?
+    fn is_invalid_sentinel(self) -> bool {
+        self == Self::invalid_sentinel()
+    }
+}
+impl InvalidSentinel for Reg {
+    fn invalid_sentinel() -> Self {
+        Reg::invalid()
+    }
+}
+impl InvalidSentinel for VirtualReg {
+    fn invalid_sentinel() -> Self {
+        VirtualReg::invalid()
+    }
+}
+impl InvalidSentinel for RealReg {
+    fn invalid_sentinel() -> Self {
+        RealReg::invalid()
+    }
+}
+impl InvalidSentinel for Writable<Reg> {
+    fn invalid_sentinel() -> Self {
+        Writable::from_reg(Reg::invalid_sentinel())
+    }
+}
+
+impl<R: Clone + Copy + Debug + PartialEq + Eq + InvalidSentinel> ValueRegs<R> {
+    /// Is this Value-to-Reg mapping valid?
+    pub fn is_valid(self) -> bool {
+        !self.parts[0].is_invalid_sentinel()
+    }
+    /// Is this Value-to-Reg mapping invalid?
+    pub fn is_invalid(self) -> bool {
+        self.parts[0].is_invalid_sentinel()
+    }
+
+    /// Return the single register used for this value, if any.
+    pub fn only_reg(self) -> Option<R> {
+        if self.len() == 1 {
+            Some(self.parts[0])
+        } else {
+            None
+        }
+    }
+
+    /// Return an iterator over the registers storing this value.
+    pub fn regs(&self) -> &[R] {
+        &self.parts[0..self.len()]
+    }
+}
+
+#[cfg(feature = "arm32")]
+impl<R: Clone + Copy + Debug + PartialEq + Eq + InvalidSentinel> ValueRegs<R> {
+    /// Create an invalid Value-in-Reg.
+    pub fn invalid() -> Self {
+        ValueRegs {
+            parts: [R::invalid_sentinel(); 4],
+        }
+    }
+    /// Create a Value-in-R location for a value stored in one register.
+    pub fn one(reg: R) -> Self {
+        ValueRegs {
+            parts: [
+                reg,
+                R::invalid_sentinel(),
+                R::invalid_sentinel(),
+                R::invalid_sentinel(),
+            ],
+        }
+    }
+    /// Create a Value-in-R location for a value stored in two registers.
+    pub fn two(r1: R, r2: R) -> Self {
+        ValueRegs {
+            parts: [r1, r2, R::invalid_sentinel(), R::invalid_sentinel()],
+        }
+    }
+    /// Create a Value-in-R location for a value stored in four registers.
+    pub fn four(r1: R, r2: R, r3: R, r4: R) -> Self {
+        ValueRegs {
+            parts: [r1, r2, r3, r4],
+        }
+    }
+
+    /// Return the number of registers used.
+    pub fn len(self) -> usize {
+        // If rustc/LLVM is smart enough, this might even be vectorized...
+        (self.parts[0] != R::invalid_sentinel()) as usize
+            + (self.parts[1] != R::invalid_sentinel()) as usize
+            + (self.parts[2] != R::invalid_sentinel()) as usize
+            + (self.parts[3] != R::invalid_sentinel()) as usize
```

Comment:
For the 4 element case:

```rust
pub struct ValueRegs {
    parts: [u32; 4],
}

impl ValueRegs {
    /// Return the number of registers used.
    pub fn len(self) -> usize {
        // If rustc/LLVM is smart enough, this might even be vectorized...
        (self.parts[0] != !0) as usize
            + (self.parts[1] != !0) as usize
            + (self.parts[2] != !0) as usize
            + (self.parts[3] != !0) as usize
    }
}
```

```asm
.LCPI0_0:
	.quad	1
	.quad	1

playground::ValueRegs::len:
	movdqu	(%rdi), %xmm0
	pcmpeqd	%xmm1, %xmm1
	pcmpeqd	%xmm1, %xmm0
	pxor	%xmm1, %xmm0
	pshufd	$246, %xmm0, %xmm1
	movdqa	.LCPI0_0(%rip), %xmm2
	pand	%xmm2, %xmm1
	pshufd	$212, %xmm0, %xmm0
	pand	%xmm2, %xmm0
	paddq	%xmm1, %xmm0
	pshufd	$78, %xmm0, %xmm1
	paddq	%xmm0, %xmm1
	movq	%xmm1, %rax
	retq
```

For the 2 element case:

```rust
pub struct ValueRegs {
    parts: [u32; 2],
}

impl ValueRegs {
    /// Return the number of registers used.
    pub fn len(self) -> usize {
        // If rustc/LLVM is smart enough, this might even be vectorized...
        (self.parts[0] != !0) as usize
            + (self.parts[1] != !0) as usize
    }
}
```

```asm
playground::ValueRegs::len:
	xorl	%eax, %eax
	cmpl	$-1, %edi
	setne	%al
	movabsq	$-4294967296, %rcx
	cmpq	%rcx, %rdi
	adcq	$0, %rax
	retq
```

Reply:
Even better, for n=2 everything is in registers; scalar replacement of aggregates FTW \o/

Reply:
Thanks @bjorn3 for showing the assembly here. I think that, yes, we want to avoid control flow and let the compiler use its cleverness here.